File: Tree/Trie.java
class TrieNode {
    // Initialize your data structure here.
    boolean word;
    TrieNode[] children;
    public TrieNode() {
        this.word = false;
        this.children = new TrieNode[26];
    }
}

public class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    // Inserts a word into the trie.
    public void insert(String word) {
        TrieNode node = root;
        for(char c:word.toCharArray()){
            if(node.children[c-'a']==null)
                node.children[c-'a'] = new TrieNode();
            node = node.children[c-'a'];
        }
        node.word = true;
    }

    // Returns if the word is in the trie.
    public boolean search(String word) {
        TrieNode node = root;
        for(char c:word.toCharArray()){
            if(node.children[c-'a']==null)
                return false;
            node = node.children[c-'a'];
        }
        return node.word;
    }

    // Returns if there is any word in the trie
    // that starts with the given prefix.
    public boolean startsWith(String prefix) {
        TrieNode cur = root;
        for(char c:prefix.toCharArray()){
            if(cur.children[c-'a']==null)
                return false;
            cur = cur.children[c-'a'];
        }
        return true;
    }
}

File: Tree/Trie.py
class TrieNode(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.word = False;  #indicate if there is a word end in this node
        self.children = {}; #HashMap: record all its children
        

class Trie(object):

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        node = self.root;
        for w in word:
            if w not in node.children:
                node.children[w] = TrieNode()
            node = node.children[w]
        node.word = True
        

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        node = self.root
        for w in word:
            if w not in node.children:
                return False
            node = node.children[w]
        return node.word
        

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie
        that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        node = self.root
        for w in prefix:
            if w not in node.children:
                return False
            node = node.children[w]
        return True
            
        
